targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/serial_api.c

```diff
@@ -0,0 +1,378 @@
+                                        (1<<CMSDK_GPIO_ALTFUNC_UART0_TX_SET);
+#endif
+          }
+          if((int)uart_rx != NC)
+          {
+            CMSDK_UART0->CTRL |= 0x02;  // RX enable
+#if CMSDK_GPIO_ALTFUNC_UART0_RX
+            GPIO_MAP[CMSDK_GPIO_UART0_RX_GPIO_NUM]->ALTFUNCSET |=
+                                        (1<<CMSDK_GPIO_ALTFUNC_UART0_RX_SET);
+#endif
+          }
+        }
+        break;
+        case UART_3:     //Sheild1 UART
+        {
+          CMSDK_UART2->CTRL = 0x00;    // Disable UART when changing configuration
+          if((int)tx != NC)
+          {
+            CMSDK_UART2->CTRL = 0x1;   // TX enable
+            GPIO_MAP[CMSDK_GPIO_SH0_UART2_TX_GPIO_NUM]->ALTFUNCSET |=
+                                    (1<<CMSDK_GPIO_ALTFUNC_SH0_UART2_TX_SET);
+          }
+          if((int)rx != NC)
+          {
+            CMSDK_UART2->CTRL |= 0x2;  // RX enable
+            GPIO_MAP[CMSDK_GPIO_SH0_UART2_RX_GPIO_NUM]->ALTFUNCSET |=
+                                    (1<<CMSDK_GPIO_ALTFUNC_SH0_UART2_RX_SET);
+          }
+        }
+        break;
+    }
+
+    // set default baud rate and format
+    serial_baud  (obj, 9600);
+
+    // pinout the chosen uart
+    pinmap_pinout(tx, PinMap_UART_TX);
+    pinmap_pinout(rx, PinMap_UART_RX);
+
+    switch (uart) {
+        case UART_0: obj->index = 0; break;
+        case UART_1: obj->index = 1; break;
+        case UART_2: obj->index = 2; break;
+        case UART_3: obj->index = 3; break;
+    }
+    uart_data[obj->index].sw_rts.pin = NC;
+    uart_data[obj->index].sw_cts.pin = NC;
+    serial_set_flow_control(obj, FlowControlNone, NC, NC);
+
+    is_stdio_uart = (uart == STDIO_UART) ? (1) : (0);
+
+    if (is_stdio_uart) {
+        stdio_uart_inited = 1;
+        memcpy(&stdio_uart, obj, sizeof(serial_t));
+    }
+}
+
+void serial_free(serial_t *obj) {
+      uart_data[obj->index].serial_irq_id = 0;
+}
+
+// serial_baud
+// set the baud rate, taking in to account the current SystemFrequency
+void serial_baud(serial_t *obj, int baudrate) {
+    // The MPS2 has a simple divider to control the baud rate. The formula is:
+    //
+    // Baudrate = PCLK / BAUDDIV
+        //
+        // PCLK = 24 Mhz
+        // so for a desired baud rate of 9600
+        //  24000000 / 9600 = 2500
+    //
+    //check to see if minimum baud value entered
+    int baudrate_div = 0;
+    baudrate_div = SystemCoreClock / baudrate;
+    if(baudrate >= 16){
+    switch ((int)obj->uart) {
+        case UART_0: CMSDK_UART1->BAUDDIV = baudrate_div; break;
+        case UART_1: CMSDK_UART3->BAUDDIV = baudrate_div; break;
+        case UART_2: CMSDK_UART0->BAUDDIV = baudrate_div; break;
+        case UART_3: CMSDK_UART2->BAUDDIV = baudrate_div; break;
+        default: error("serial_baud"); break;
+    }
+    } else {
+        error("serial_baud");
+    }
+
+}
+
+void serial_format(serial_t *obj, int data_bits, SerialParity parity, int stop_bits) {
```

Comment:
@bridadan  The CMSDK APB UART is a simple design that supports 8-bit communication without parity, and is fixed at one stop bit per configuration. Ref. DDI0479C_cortex_m_system_design_kit_r1p0_trm.pdf. However, one comment and mbed error will be added in this function.